Tutorials: Add macOS install and Test Net connect to Install rippled

jhaaaa opened the pull request:
- Update Install `rippled` tutorial to include steps for installing `rippled` on macOS
- Add info about how to connect `rippled` to the XRP Test Net
- Incorporate feedback from PM
- Base on install steps from QA
- Base on https://github.com/ripple/rippled/tree/develop/Builds/macos

jhaaaa committed:
just getting started

jhaaaa committed:
just starting out

jhaaaa committed:
first pass for review

jhaaaa committed:
first pass for review

jhaaaa committed:
fix links

jhaaaa commented:
Hi @manojsdoshi - if you have some time, it would be great to get your input on this tutorial about `rippled` install for macOS. I based the flow on your testing steps and on feedback from PM. Thank you for any feedback you have time to provide!

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+{{n.next()}}. Install [Xcode](https://developer.apple.com/download/).
 
-4. Verify the signature on the `rippled` software package:
+{{n.next()}}. Install Xcode command line tools. ***TODO: Comment from PM: Installing homebrew removes the need to install xcode command line tools separately. JHA: homebrew lists xcode command line tools as a requirement: https://docs.brew.sh/Installation#requirements. Do we keep or remove this step?***
```
Ah, maybe its dependent on some sort of installation settings where you can install both simultaneously. I think its safer bet to leave it as for now, I can look into it.

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+        brew update
 
-7. Start the `rippled` service
+{{n.next()}}. Use Homebrew to install dependencies. ***TODO: Added boost per PM's request: Installing Boost directly is going to frustrate people - brew can do it in one command. I ended up installing Boost with Brew with no issues. Which release to choose? JHA: I did `brew install boost` and brew installs 1.67.0 in /usr/local/Cellar/boost/1.67.0_1 -- is this okay? If we install Boost via brew, d ...
```
(1) Using homebrew removes the need to compile it from scratch. You also do not need to set environment variables yourself. The only catch is brew updates to the latest version of boost, which the Rippled team must first OK in order for it to work. 1.68 is the latest release and it works fine, but 1.69 will need to be checked. We may need to periodically update to reflect this?

(2) Not sure why it didn't work, I did not try to debug it and this is why I tried to use homebrew instead.

mDuo13 replied:
Users will have to compile boost themselves any time the version of boost that's in the homebrew repositories is not a version that's supported by rippled. At the moment the two are compatible but I'm not convinced that's going to remain true for the long term. Perhaps we should have both instructions available and switch the recommendations periodically based on whether the easier brew install is compatible.

mDuo13 replied:
And as Daniel alluded, the compatibility between the Brew and rippled versions can break in two ways:

- brew updates Boost to a new version that's not compatible with the current release of rippled
- rippled updates to a new version that's not compatible with the version of Boost in the Brew repositories

The former is more likely than the latter since Brew tends to update quickly, but the latter is also possible if the person who's currently maintaining the Boost package for Brew changes or takes a long vacation or something, or if the new version of Boost is held up from being added to Brew because of some other incompatibilities with other Brew software that requires it. (Similar situations have happened to me on Arch Linux.)

The "compile it yourself" instructions should be fairly stable and will only change as the result of `rippled` updates, so that would be a much easier maintenance path going forward, even though it's more work for users. The "use the version from Brew" instructions are only sometimes going to be possible and if we want to support those we'll have to keep track of when new versions of Boost are added to Brew. When the Brew-based installation instructions work, they are much easier for users, so I leave it to @jhaaaa to decide whether that's worth the effort of keeping track.

garay-daniel replied:
I guess we should decide how tricky boost compilation is. Im not sure how often Boost updates, but thats about how often we would need to update these instructions with the new version for brew commands. You can have brew install a specific version of a formula, which is the 3rd option as well. Something like brew install boost 1.67.0 ...

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        brew install git cmake pkg-config protobuf openssl ninja boost
+
+{{n.next()}}. Set the `BOOST_ROOT` environment variable to point to the new directory created by the Homebrew installation of Boost. Put this environment variable in your `.bash_profile` so it's automatically set when you log in. For example, add the following line to the file. To find your Boost install directory, use `brew info boost`. ***TODO: Feedback from PM: Have to deal with environment va ...
```
If you install with homebrew you do not need to set your own environment variables. I think just mentioning its bash_profile would be fine. You can use your .bashrc, but it requires also using your bash_profile as far as I know. I just thought "rc files" could be misinterpreted.

jhaaaa replied:
Got it. When I installed, it did not update my BOOST_ROOT env var, but that is probably because it was already set? Maybe it doesn't overwrite/update if a value exists? I've updated the text to address - let's review and see if it works for you.

mDuo13 replied:
I'm not sure on this, but maybe you only need to set the BOOST_ROOT environment variable if you're compiling Boost yourself. If you're able to install a compatible version of Boost via Brew (per the debate above) then the part about setting your BOOST_ROOT env var may be moot.

garay-daniel replied:
Sorry if I was not clear. You definitely must set the env variable if compiling yourself. If you also install a brew version, you would have 2 boost(s) since brew uses its directory (different from yours probably) to install. A fresh install with only brew would not need the BOOST_ROOT env variable.

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        brew install git cmake pkg-config protobuf openssl ninja boost
+
+{{n.next()}}. Set the `BOOST_ROOT` environment variable to point to the new directory created by the Homebrew installation of Boost. Put this environment variable in your `.bash_profile` so it's automatically set when you log in. For example, add the following line to the file. To find your Boost install directory, use `brew info boost`. ***TODO: Feedback from PM: Have to deal with environment va ...
```
You do not need to set the env variables with brew, but if you choose to install in another way you will have to do it. I think bash_profile is clear enough. I think you can use your .bashrc as well, but it requires putting something into your bash_profile too as far as I know.

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        source .bashrc
+
+{{n.next()}}. Clone the `rippled` source code into your desired location and access the `rippled` directory.  ***TODO: Note from PM: Cloning github repo is going to require proper github setup. Should we provide some guidance or references? JHA: What is the proper github setup? I may have set this up a whlie ago and I can't remember the steps that may have been required.***
```
Leave it up to you here, I imagine most will have proper github setup, but was worth mentioning. Maybe a simple line to make that clear is enough? You need a github acc, setup your ssh key, etc. 

https://help.github.com/articles/set-up-git/

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        cmake --build . -- -j 4
+
+      This example uses a `-j` parameter set to `4`, which uses four processes to build in parallel. The best number of processes to use depends on how many CPU cores your hardware has available. Use `sysctl -a | grep machdep.cpu | grep core_count` to get your CPU core count. ***TODO: the mac build page in the github repo goes much more in-depth about cmake build options - should we include these ...
```
I don't think its necessary and did not use it myself. Not sure what the background for those instructions are however, maybe Rome knows more?

jhaaaa replied:
Sounds great - let's see if Manoj and Rome have feedback. I'm going to add Rome as a reviewer once I get a first round of reviews incorporated. 👍

mDuo13 replied:
When I updated the Ubuntu equivalent of these instructions, I didn't want to overwhelm the reader with options and possibilities, so I left a couple of tips but didn't go into extreme detail. I also don't like relying on a series of several piped commands including a `grep` for getting your CPU count, since a minor change in the OS's output format could break that. I went with a slightly more generic message about where to get information about your processor.

These instructions can probably be based on the changes I made there: https://github.com/ripple/ripple-dev-portal/blame/master/content/tutorials/manage-the-rippled-server/build-run-rippled-ubuntu.md#L107

jhaaaa replied:
Great - I will leave out the reference to the in-depth cmake build options. 👍 

@mDuo13 - Regarding the grep - I added the command because `cat /proc/cpuinfo` didn't work for me in macOS. I see your point about it being brittle! Is there a more stable way to get this info for macOS? Or perhaps I should just leave the command out for macOS?

mDuo13 replied:
Does this command work on macOS?

```
 sysctl -n hw.ncpu 
```

jhaaaa replied:
YES! 🙌 Thank you!

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        ./rippled --unittest
+
+{{n.next()}}. Create a copy of the example config file and save it to a location that enables you to run `rippled` as a non-root user (recommended). ***TODO: unlike the centos and ubuntu installs on this page, the macos install requires that you manually put rippled.cfg and validator.txt in place. At least that is what I experienced - true?***
```
Yes I had to create both.

garay-daniel commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +81,193 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+      This example uses a `-j` parameter set to `4`, which uses four processes to build in parallel. The best number of processes to use depends on how many CPU cores your hardware has available. Use `sysctl -a | grep machdep.cpu | grep core_count` to get your CPU core count. ***TODO: the mac build page in the github repo goes much more in-depth about cmake build options - should we include these ...
+
+{{n.next()}}. Run unit tests built into the server executable. This could take about 5 minutes, depending on your hardware specs. (optional, but recommended) ***TODO: We should provide info about what to do if you get failures. What should the user do?***
```
Yes agreed, we need to learn what those failures are. I can ask.

jhaaaa replied:
Fantastic - thank you!

jhaaaa committed:
PM feedback

jhaaaa committed:
feedback

jhaaaa committed:
feedback

jhaaaa commented:
@garay-daniel and @mDuo13 - wow, thank you so much for your thoughtful and detailed feedback on this PR! 👍 I think I've addressed most of our open items with this latest push. Please take a look to ensure that I've met your requirements. You can search on `TODO` to find my comments/questions.

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -26,6 +25,44 @@ Amazon EC2's `m3.large` VM size may be appropriate depending on your workload. A
 
+## Installation on Ubuntu with alien
+
+<!--{# ***TODO: this doc states that ubuntu has received the highest level of QA and testing - so I moved it above centos and macOS install sections. Okay?*** #}-->
```
I think this is fair since most of the development and testing happens on Ubuntu.

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+That said, macOS is suitable for many development and testing tasks.
 
-        $ sudo rpm -Uvh https://mirrors.ripple.com/ripple-repo-el7.rpm
+This section assumes that you are using macOS X 10.0 or higher. ***TODO: okay?***
```
This makes sense since most people who are developing at Ripple are on macOS X 10.0 (most people are on macOS High Sierra – 10.13). 
Although we haven't upgraded to the latest macOS - "10.14 Mojave" so I am not sure if anything might have changed in that version. Not sure how to handle this in the docs maybe put a note stating that we have tested only till 10.13

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        brew install git cmake pkg-config protobuf openssl ninja
+
+{{n.next()}}. Install a supported version of Boost. Supported versions include Boost 1.67.x - 1.68.x. ***TODO: correctly stated supported versions? In the rippled repo, we say "Boost 1.67 or later is required." <-- Based on our discussion in this doc, I don't think this open-ended "later" part is true, is it?***
```
With the current release, we have tested with boost 1.67 and support that. 
We can put a note that, it will work with 1.68 but not yet tested.

jhaaaa replied:
I'm going to leave out the mention of 1.68. When I put it in the doc - it seems to scream out - install me. 😹

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
-It can take several minutes for `rippled` to sync with the rest of the network, during which time it outputs warnings about missing ledgers. After that, you have a fully functional stock `rippled` server that you can use for local signing and API access to the XRP Ledger.
+It can take several minutes for `rippled` to sync with the rest of the network, during which time it outputs warnings about missing ledgers.
+
+Here's an excerpt of what you can expect to see in your terminal: ***TODO: you'll see this in the CLI for the macOS install because we tell them to start the service using `sudo ./rippled`, but not for centos and ubuntu. For centos and ubuntu, we tell the user to start the service using `sudo systemctl start rippled.service`, which starts the service, but doesn't display a log of the service star ...
```
For centos and ubuntu after running 'sudo systemctl start rippled' the user can run another command  'sudo systemctl status rippled' to verify is rippled has started.

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+    Pick one of the following Boost installation methods: ***TODO: I provided the two options below because these are the most straightforward ways for a user to install one of the two supported versions of Boost. In the future, if there are more supported versions or more brew boost formalae, there may be more options to document.***
+
+    * To install Boost 1.67.0, use `brew install boost@1.67`. ***TODO: When Boost 1.67.0 is no longer supported, we'll need to change this formula. Until then, the user will be able to use this command to get a supported version of Boost. `brew install boost` also installs 1.67.0, but as discussed, we don't want to provide the command because it won't always install 1.67.0 and may cause the user  ...
```
lgtm

jhaaaa commented:
@mDuo13 and @garay-daniel - Manoj helped get Nik's confirmation that we'll go with Boost 1.67.0 only for now. I'll update the doc to reflect this and send a note when this PR is ready for another review.

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        source .bash_profile
+
+{{n.next()}}. Clone the `rippled` source code into your desired location and access the `rippled` directory. To do this, you'll need to set up Git (installed earlier using Homebrew) and GitHub. For example, you'll need to create a GitHub account and set up your SSH key. For more information, see [Set up git](https://help.github.com/articles/set-up-git/). ***TODO: does this sound okay for introduc ...
```
lgtm

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +83,222 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        git checkout release
+
+{{n.next()}}. In the `rippled` directory you just cloned, create your build directory and access it. ***TODO: Does this "In the rippled directory you just cloned" language make sense? I just want to tell the user where to create the build directory.*** For example:
```
lgtm

jhaaaa committed:
manoj's feedback

jhaaaa commented:
@mDuo13, @garay-daniel, and @manojsdoshi: Here's and updated version. So close! Please search on TODO to see my outstanding comments/questions. Any feedback/guidance you have time to provide is greatly appreciated!

jhaaaa committed:
added test net connect info

jhaaaa commented:
@mDuo13, @garay-daniel, and @manojsdoshi - added info about how to connect your `rippled` to the XRP Test Net. Add to the Additional Configuration section. Just a few more steps to review. Thank you!

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -295,9 +265,9 @@ Once you have your stock `rippled` server running, you may want to consider runn
 `rippled` should connect to the XRP Ledger with the default configurations. However, you can change your settings by editing the `rippled.cfg` file. For recommendations about configuration settings, see [Capacity Planning](capacity-planning.html).
 
-Changes to the `[node_db]`, `[debug_logfile]`, or `[database_path]` sections may require you to give the `rippled` user and group ownership to your new configured path: ***TODO: I added [node_db] to this list - okay? When I try the following command, `chown -R rippled:rippled /var/lib/rippled/db/rocksdb`, I get `chown: rippled: illegal group name.` Is `rippled:rippled` just an example below -- or ...
+Changes to the `[node_db]`, `[debug_logfile]`, or `[database_path]` sections may require you to give the `rippled` user and group ownership to your new configured path: ***TODO: I added [node_db] to this list - okay?***
```
lgtm

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -295,9 +265,9 @@ Once you have your stock `rippled` server running, you may want to consider runn
+Changes to the `[node_db]`, `[debug_logfile]`, or `[database_path]` sections may require you to give the `rippled` user and group ownership to your new configured path: ***TODO: I added [node_db] to this list - okay?***
 
-    $ chown -R rippled:rippled <configured path>
+    $ chown -R rippled:rippled <configured path> ***TODO: Is `rippled:rippled` just an example value? When I run this command on macOS, I get `chown: rippled: illegal group name.` When I run it on centos, I get chown: changing ownership of ... : Operation not permitted***
```
This group doesn't exist on your Mac that's why you get this error. 
When you spin up aws machines to test centos/Ubuntu (I assume that's what you are doing) this user group is created by default (OPS has this configured) hence you don't get this error. 

But you have the right idea to give permissions to the user who is going to run rippled (maybe we don't need to give an example of how to change permissions? but just say we need to get the right permission to start rippled)

mDuo13 replied:
Options:

1. Tell the user to run rippled as their own user. This is a reasonable balance of safe and easy—not as safe as running as a dedicated user like `rippled`, not as easy as running as root, but in between. Change the example command to use `$USER` (the environment variable for the current user, assuming that's also valid on macOS like it is on Linux).
2. Tell the user to create a `rippled` user for safety and provide steps for (a) how to create such a user, (b) how to give it the right permissions, and (c) how to run `rippled` as that user. This is fairly standard system admin-level stuff, which may or may not be a bit too much work for the intended audience here. We're already telling them not to run on macOS in production so taking production-level precautions is probably too much here.

I'd lean towards 1 since it's more expedient. There are third-party docs they can leverage if they want to re-create a production-like environment on macOS despite our recommendations, and we probably shouldn't provide recommendations for things we don't support anyway.

manojsdoshi commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -201,7 +173,7 @@ This section assumes that you are using macOS X 10.0 or higher. ***TODO: okay?**
         cp cfg/rippled-example.cfg /etc/opt/ripple/rippled.cfg
 
-{{n.next()}}. Edit `rippled.cfg` to set necessary file paths. The user you plan to run `rippled` as must have write permissions to all of the paths you specify here. ***TODO: I think we need to add this step to the centos and ubuntu tasks, correct? For centos and ubuntu, the config files are automatically placed in the correct directories, but the following paths still need to be manually updated ...
+{{n.next()}}. Edit `rippled.cfg` to set necessary file paths. The user you plan to run `rippled` as must have write permissions to all of the paths you specify here. ***TODO: We need to add this step to the centos and ubuntu tasks, correct? For centos and ubuntu, the config files are automatically placed in the correct directories, but the following paths still need to be manually checked/updated ...
```
Yup, it's a good idea for the user to verify if they have the correct permissions rather than rippled failing to start and the user opening a support ticket for us.

manojsdoshi replied:
Maybe this can go as a tip to the user or debug steps if they cant start rippled

mDuo13 replied:
Let's:

1. Keep the scope of this task from growing to modify the other installs
2. Not add steps telling people to check / edit things they don't need to edit (if they're following the directions in the intended scenario).
    - To be clear, we may need to add these steps to the macOS build & run instructions, but let's not add them to the CentOS / Ubuntu installs from packages.
3. Instead of adding steps, let's just put in a brief note with a link to [the existing docs for rippled not being able to read a config file](https://developers.ripple.com/server-wont-start.html#failed-to-open-etcoptripplerippledcfg). 
4. Improve those directions in a later ticket if necessary.

manojsdoshi reviewed (approved):
LGTM

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+{% set n = cycler(* range(1,99)) %}
 
-4. Verify the signature on the `rippled` software package:
+{{n.next()}}. Install [Xcode](https://developer.apple.com/download/).
```
I don't remember what the outcome is from our previous discussion on this topic, but I still think it's better to follow the more concise paradigm of having all steps (after the first) be numbered "0" if you want to have list steps be automatically numbered.

After all, this:

```
1. step 1
0. step 2
0. step 3
```

... is much easier to read and write than this:

```
{% set n = cycler(* range(1,99)) %}

{{n.next()}}. Step 1
{{n.next()}}. Step 2
{{n.next()}}. Step 3
```

... and the former _also_ has the advantage of working in the GitHub preview and other Markdown parsers that don't use Dactyl's pre-parser.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+        brew install git cmake pkg-config protobuf openssl ninja
+
+{{n.next()}}. Install Boost 1.67.0. Currently, we support Boost 1.67 only.
```
Nit: "We" don't support Boost; `rippled` 1.1.x _is compatible with_ Boost 1.67.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+{{n.next()}}. Ensure that your `BOOST_ROOT` environment variable points to the directory created by the Boost installation. To find your Boost install directory, use `brew info boost`. Put this environment variable in your `.bash_profile` file so it's automatically set when you log in. For example:
+
+        export BOOST_ROOT=/usr/local/Cellar/boost/1.67.0_1
```
An oddity (which applies to the other version of these instructions) worth note: This is the only code sample in these steps which isn't (intended as) a commandline instruction. Almost coincidentally, it works as one, but it won't have quite the intended effect—it's temporary only, and the next step is pointless if you don't put this in your `.bash_profile`.

Our examples are somewhat inconsistent about this, but one commonly used convention is to start commandline examples with a `$` prompt (for non-administrative commands) or with a `#` prompt (when the commandline has root-level—not just sudo—permissions; rare). There are some downsides (you have to remove the $ when you copy-paste commands) but this is one case where the difference would help convey some useful information. We should discuss making a note about this in our style guide and being consistent about it.

jhaaaa replied:
Thank you so much for taking the time to explain this convention to me. 👍  My work life is changed forever. I've updated this doc accordingly. It would be great to include it in the style guide.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+{{n.next()}}. `rippled` requires the `rippled.cfg` configuration file to run. You can find an example config file, `rippled-example.cfg` in `rippled/cfg`. Make a copy and save it as `rippled.cfg` in a location that enables you to run `rippled` as a non-root user. Access the `rippled` directory and run:
+
+        mkdir -p /etc/opt/ripple ***TODO: changed this from ~/.config/ripple to match guidance coming in conf-file-location.md PR -- okay?***
```
Seems fine (assuming this works on macOS; I haven't tested that). You _can_ have the config file in your home dir and owned by a non-admin user, which has pros and cons and relates to what user you plan to run `rippled` as.

- Running as root / admin is easy to set up, since you probably don't have to worry about wrong permissions
    - If you're running `rippled` as a root user, it won't look in a non-root user's home dir, so that won't work.
- Running as not-root lets you limit the damage the server can do if something unexpectedly bad happens, like going berserk due to a bug or being maliciously controlled by some external person. We do our best to make these kinds of things almost impossible, but it's a sensible (and common) precaution to take when running any server that's connected to the internet at large.
    - Typically Linux  packages for servers are configured to automate the work to add a user specifically for that server, then give it permissions to _only_ the files it wants. This probably includes read-only permissions to its config file (which are owned by and writable only as root)
    - For dev purposes, you can run the server as your own user, which is slightly less setup but still more secure than running as root. This last case is the one where you want the config file in a home directory.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+{{n.next()}}. `rippled` requires the `validators.txt` file to run. You can find an example validators file, `validators-example.txt`, in `rippled/cfg/`. Make a copy and save it as `validators.txt` in the same folder as your `rippled.cfg` file. Access the `rippled` directory and run:
+
+        cp cfg/validators-example.txt /etc/opt/ripple ***TODO: changed this from ~/.config/ripple to match guidance coming in conf-file-location.md PR -- okay?***
```
This needs to change the filename of the file when it moves it, to remove the `-example` part:

```
cp cfg/validators-example.txt /etc/opt/ripple/validators.txt
```

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+{{n.next()}}. Access your build directory, `my_build` for example, and start the `rippled` service.
+
+        sudo ./rippled
```
This runs `rippled` as root—not recommended. If you set up file paths for your config and databases to directories you can write, then `rippled` should run without root permissions.

Caution: if you've ever started rippled as root (including using `sudo`) then it probably created some database files as a root user, which means it'll fail with a permissions error if you try to run it as non-root user, _even if the user has permissions to the directories containing the databases_. To fix this, you have to (using root permissions) delete `rippled`'s database files (in particular, the ones it created when it ran as an admin user, but also possibly ones it created when run as a non-root user before that, because they may be incomplete/corrupt if you delete some database files). The easiest way to do this is `sudo rm -r` the `rippled` database directory (assuming you don't have anything else in that directory) and then re-create it without sudo / admin, as the user you intend to run `rippled` with.

jhaaaa replied:
Do you have a 🔮 in which you can see what I'm doing on my command line? 😂  Deleted my db files and it ran without sudo. Removed `sudo` here.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
-It can take several minutes for `rippled` to sync with the rest of the network, during which time it outputs warnings about missing ledgers. After that, you have a fully functional stock `rippled` server that you can use for local signing and API access to the XRP Ledger.
+It can take several minutes for `rippled` to sync with the rest of the network, during which time it outputs warnings about missing ledgers.
+
+<!--{# ***TODO: for the future, provide a reference for what the most common and unintuitive messages mean -- esp when signaling possible errors.*** #}-->
```
Link this page: https://developers.ripple.com/understanding-log-messages.html

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+    $ chown -R rippled:rippled <configured path> ***TODO: Is `rippled:rippled` just an example value? When I run this command on macOS, I get `chown: rippled: illegal group name.` When I run it on centos, I get chown: changing ownership of ... : Operation not permitted***
+
+Restart `rippled` for any configuration changes to take effect: ***TODO: when I try this on macOS, I get `sudo: service: command not found`. I don't think the `service` command exists for macOS. For macOS, can the user just stop (ctrl+c? Not sure if that is the proper way to stop the service) and start rippled (sudo ./rippled) instead? For centos and ubuntu, we use `sudo systemctl start rippled.s ...
```
The `service` command is part of systemd, which is not part of macOS. Also, running `./rippled` or `sudo ./rippled` is not running it as a service.

So yeah, Ctrl-C would be one way to stop it if you did that. 

 I'm not sure what the macOS equivalent of "setting up a systemd unit to run it as a service" is.

jhaaaa replied:
Great to know that issuing the `./rippled` command is not running `rippled` as a service. 😅 For now, I'll go with Ctrl-C to stop and `./rippled` to start it up again. For the future, I will look into a command for starting `rippled` as a service for macOS.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+
+### Connect Your `rippled` to the XRP Test Net
+
+Ripple has created the [XRP Test Network](https://ripple.com/build/xrp-test-net/) to provide a testing platform for the XRP Ledger. XRP Test Net funds are not real funds and are intended for testing only. You may want to connect your `rippled` server to the XRP Test Net to test out and understand `rippled` functionality before connecting to the production XRP Ledger Network. ***TODO: stated corre ...
```
:+1: 

It's also about testing out that your own code interacts correctly with `rippled`.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+3. Restart `rippled`.
 
-Restart `rippled` for any configuration changes to take effect:
+4. To verify that your `rippled` is connected to the XRP Test Net, go to the [XRP Test Net Faucet](https://developers.ripple.com/xrp-test-net-faucet.html) and click **Generate credentials**. Make a note of the **Address** value. Make the [`account_info`](https://developers.ripple.com/account_info.html) request to your `rippled` server. ***TODO: Any ideas for a better way to this check?***
```
This works. You can also run `server_info` and see what ledger version your server is on (assuming the state is `full`/`proposing` and it's not still syncing). If it's close to or the same as what `s1.ripple.com`'s server_info reports, then you're on the production network. If it's much smaller (like, fewer digits) then you're on the testnet.

mDuo13 commented on content/tutorials/manage-the-rippled-server/install-rippled.md:
```diff
@@ -46,59 +85,234 @@ This section assumes that you are using CentOS 7 or Red Hat Enterprise Linux 7.
+4. To verify that your `rippled` is connected to the XRP Test Net, go to the [XRP Test Net Faucet](https://developers.ripple.com/xrp-test-net-faucet.html) and click **Generate credentials**. Make a note of the **Address** value. Make the [`account_info`](https://developers.ripple.com/account_info.html) request to your `rippled` server. ***TODO: Any ideas for a better way to this check?***
 
-        $ sudo service rippled restart
+      If the request returns account information, your `rippled` is connected to the XRP Test Net. If the request cannot find the account, your `rippled` is not connected to the XRP Test Net. Check your `rippled.cfg` and `validators.txt` configurations.
```
Technically this could be fooled if someone funded your testnet credentials on prod. It's not likely to happen in the timescale described here but if you tried this same test much later after creating your testnet credentials, it could happen.

mDuo13 reviewed (changes requested):
This is shaping up nicely. I left some info and hopefully we can get this in good shape without dragging it out much more, especially since the more changes that happen here the more conflicts that have to be resolved with #471.

jhaaaa committed:
Rome, Manoj feedback

jhaaaa committed:
Rome, Manoj feedback

jhaaaa commented:
@manojsdoshi - thank you for all of your feedback and for your approval!

@mDuo13 and @garay-daniel - thank you for your feedback and guidance as well! I believe this commit addresses all comments/feedback. No more "TODO"s to search for 😂. When you have a moment - please take a look and provide your approval, if worthy!

mDuo13 reviewed (approved):
:+1: The revisions look great!

jhaaaa commented:
The build is breaking because of a broken link to https://cmake.org/cmake/help/v3.0/variable/CMAKE_BUILD_TYPE.html. The link works when I access it directly and when I access it from my locally built doc.
